### PR TITLE
Add new fields to service to track new data type and name

### LINKED ISSUE #1244: [META 621] Improve Granularity for SQL Databases
[issue details](https://github.com/elastic/apm/issues/621)

[issue details](https://github.com/elastic/apm/issues/622)

- start collecting `span.service.target.name` and `span.service.target.type` fields and deprecate `span.destination.service.resource`.
- depending on APM server version send 
  - `span.destination.service.resource` (APM server < 8.3)
  - `span.service.target.name` and `span.service.target.type` (APM server >= 8.3)
- replace the internal usage of the `span.destination.service.resource` field with the new fields (e.g. compressed spans algorithm, inferring `span.service.target.name`, `span.service.target.type` instead of `span.destination.service.resource`).

### PR DESCRIPTION
**Pull request by jaggederest**
## What does this pull request do?

Adds new fields to agent-server protocol, specifically span.context.service.target.type and span.context.service.target.name

## Why is it important?

These fields will be used in the future as the correct way to capture service details in the agent-server API

## Checklist

- [x] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
- [x] My code follows the style guidelines of this project (See `.rubocop.yml`)
- [x] I have rebased my changes on top of the latest main branch

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing [**unit** tests](https://github.com/elastic/apm-agent-ruby/blob/main/CONTRIBUTING.md#testing) pass locally with my changes

- [x] I have made corresponding changes to the documentation
- [x] I have updated [CHANGELOG.asciidoc](CHANGELOG.asciidoc)
- [x] I have updated [supported-technologies.asciidoc](docs/supported-technologies.asciidoc)
- [x] Added an API method or config option? Document in which version this will be introduced

## Related issues

Relates #1244 (partially completes)

### CODE REVIEW

File: lib/elastic_apm/span/context/service.rb
Comment (estolfo) [inline lib/elastic_apm/span/context/service.rb:45]: What is this `cloud` arg for?

**Review by estolfo — CHANGES_REQUESTED** [review]

**Review by estolfo — CHANGES_REQUESTED** [review]
We also need to add in the functionality to actually capture these values for SQL databases.

**Review by estolfo — CHANGES_REQUESTED** [review]
I think it would make sense to add tests for the database instrumentations, right?
Also, don't we have to keep the existing fields for compatibility with the UI?

**Comment by jaggederest** [comment]
> I think at this point, we are only supposed to add the new fields and not stop sending any existing fields.

Just for posterity's sake, the existing sequel spy did not send destination.service.name or destination.service.type before this PR, so they are not being sent. Had to double check that to make sure I didn't accidentally delete them.

File: lib/elastic_apm/span.rb
Comment (estolfo) [inline lib/elastic_apm/span.rb:149]:
```suggestion
        target: Span::Context::Service::Target.new(name: context.destination.service.name, type: context.destination.service.type )
```

Comment (estolfo) [reply lib/elastic_apm/span.rb:149]: I also think that this should have been caught by a test. Can we add one that would cover this part of the code?

**Review by estolfo — CHANGES_REQUESTED** [review]
I noticed a typo so proposed a correction but I also suggest adding a test that would have caught the typo.

File: lib/elastic_apm/span.rb
Comment (jaggederest) [reply lib/elastic_apm/span.rb:149]: Oh good catch, and yes, absolutely, adding a test now 👍

**Review by estolfo — APPROVED** [review]

**Comment by estolfo** [comment]
I just saw that the tests are failing but I approve the code changes once they are passing. Let's get the tests passing before merging. I'm happy to do another code review once the test failures are fixed. Thanks!

**Pull request merged**
